hbase/src/main/java/magnolify/hbase/HBaseType.scala

```diff
@@ -0,0 +1,267 @@
+      }
+    } else {
+      new Record[T] {
+        private def qualifier(prefix: String, label: String): String =
+          if (prefix == null) label else s"$prefix.$label"
+
+        override def get(xs: Map[String, Array[Byte]], k: String)(cm: CaseMapper): Value[T] = {
+          var fallback = true
+          val r = caseClass.construct { p =>
+            val q = qualifier(k, cm.map(p.label))
+            val v = p.typeclass.get(xs, q)(cm)
+            if (v.isSome) {
+              fallback = false
+            }
+            v.getOrElse(p.default)
+          }
+          // result is default if all fields are default
+          if (fallback) Value.Default(r) else Value.Some(r)
+        }
+
+        override def put(k: String, v: T)(cm: CaseMapper): Map[String, Array[Byte]] =
+          caseClass.parameters.flatMap { p =>
+            p.typeclass.put(qualifier(k, cm.map(p.label)), p.dereference(v))(cm)
+          }.toMap
+      }
+    }
+  }
+
+  @implicitNotFound("Cannot derive BigtableField for sealed trait")
+  private sealed trait Dispatchable[T]
+
+  def split[T: Dispatchable](sealedTrait: SealedTrait[Typeclass, T]): HBaseField[T] = ???
+
+  implicit def gen[T]: HBaseField[T] = macro Magnolia.gen[T]
+
+  def apply[T](implicit f: HBaseField[T]): HBaseField[T] = f
+
+  def from[T]: FromWord[T] = new FromWord
+
+  class FromWord[T] {
+    def apply[U](f: T => U)(g: U => T)(implicit hbf: Primitive[T]): Primitive[U] =
+      new Primitive[U] {
+        override def size: Option[Int] = hbf.size
+        override def fromBytes(v: Array[Byte]): U = f(hbf.fromBytes(v))
+        override def toBytes(v: U): Array[Byte] = hbf.toBytes(g(v))
+      }
+  }
+
+  private def primitive[T](
+    capacity: Int
+  )(f: ByteBuffer => T)(g: (ByteBuffer, T) => ByteBuffer): Primitive[T] = new Primitive[T] {
+    override val size: Option[Int] = Some(capacity)
+    override def fromBytes(v: Array[Byte]): T = f(ByteBuffer.wrap(v).asReadOnlyBuffer())
+    override def toBytes(v: T): Array[Byte] = g(ByteBuffer.allocate(capacity), v).array()
+  }
+
+  implicit val hbfByte: Primitive[Byte] =
+    primitive[Byte](java.lang.Byte.BYTES)(_.get)(_.put(_))
+  implicit val btChar: Primitive[Char] =
+    primitive[Char](java.lang.Character.BYTES)(_.getChar)(_.putChar(_))
+  implicit val hbfShort: Primitive[Short] =
+    primitive[Short](java.lang.Short.BYTES)(_.getShort)(_.putShort(_))
+  implicit val hbfInt: Primitive[Int] =
+    primitive[Int](java.lang.Integer.BYTES)(_.getInt)(_.putInt(_))
+  implicit val hbfLong: Primitive[Long] =
+    primitive[Long](java.lang.Long.BYTES)(_.getLong)(_.putLong(_))
+  implicit val hbfFloat: Primitive[Float] =
+    primitive[Float](java.lang.Float.BYTES)(_.getFloat)(_.putFloat(_))
+  implicit val hbfDouble: Primitive[Double] =
+    primitive[Double](java.lang.Double.BYTES)(_.getDouble)(_.putDouble(_))
+  implicit val hbfBoolean: Primitive[Boolean] =
+    from[Byte](_ == 1)(if (_) 1 else 0)
+  implicit val hbfUUID: Primitive[UUID] =
```

Comment:
nice! should we include support for java.time classes as well?

Reply:
Maybe in an unsafe package. As there could be several way to store dates (string bytes, ms long bytes, ...), we should not arbitrarily prick one here.